File: abelkhan_type_component/component/hub/hub_msg_handle.ts
import abelkhan = require('../../abelkhan_type/ts/abelkhan');
import signals = require('../../service/signals/signals');
import hub_protcol = require('../../protcol/ts/hub');

export class hubproxy{
    public ch : abelkhan.Ichannel;
    public hub_name : string;
    public hub_type : string;
    constructor(ch:abelkhan.Ichannel, hub_name: string, hub_type: string){
        this.ch = ch;
        this.hub_name = hub_name;
        this.hub_type = hub_type;
    }

}

export class hub_msg_handle{
    public signals_hubproxy : signals.signals<hubproxy>;
    private hub_call_hub_module : hub_protcol.hub_call_hub_module;
    constructor(modules:abelkhan.modulemng){
        this.signals_hubproxy = new signals.signals<hubproxy>();

        this.hub_call_hub_module = new hub_protcol.hub_call_hub_module(modules);
        this.hub_call_hub_module.cb_reg_hub = this.reg_hub.bind(this);
    }

    public reg_hub(hub_name: string, hub_type: string){
        let ch = this.hub_call_hub_module.current_ch;
        let rsp = <hub_protcol.rsp_reg_hub>(this.hub_call_hub_module.rsp);
        rsp.rsp();

        this.signals_hubproxy.emit(new hubproxy(ch, hub_name, hub_type));
    }
}

File: abelkhan_type_csharp_component_msgpack/component/hub/hub_msg_handle.cs
/*
 * hub_msg_handle
 * qianqians
 * 2020/6/4
 */

using System;

namespace abelkhan
{
    public class hub_msg_handle
    {
        private hubmanager _hubmng;
        private hub_call_hub_module _module;
        private hub_cmd_dispatcher _cmd_dispatcher;
        public hub_msg_handle(modulemng modules, hubmanager _hubmanager, hub_cmd_dispatcher dispatcher)
        {
            this._hubmng = _hubmanager;
            _cmd_dispatcher = dispatcher;
            _module = new hub_call_hub_module(modules);
            _module.onreg_hub += reg_hub;
            _module.onreq_hub_cmd += do_hub_cmd;
        }

        public event Action<hubproxy> on_hubproxy;
        public void reg_hub(string hub_name, string hub_type)
        {
            var ch = _module.current_ch;
            var rsp = (rsp_reg_hub)(_module.rsp);
            rsp.rsp();

            hubproxy _hubproxy = _hubmng.reg_hub(ch, hub_type, hub_name);
            if (on_hubproxy != null)
            {
                on_hubproxy(_hubproxy);
            }
        }

        private async void do_hub_cmd(string cmd, string param)
        {
            var rsp = (rsp_req_hub_cmd)(_module.rsp);
            string respStr = await _cmd_dispatcher.Dispatch(cmd, param);
            rsp.rsp(respStr);
        }
    }
}